cmd/govim/signs.go

```diff
@@ -0,0 +1,148 @@
+	argDict := struct {
+		Text          string `json:"text"`
+		TextHighlight string `json:"texthl"`
+	}{errorSign.text, errorSign.textHighlight}
+
+	if v.ParseInt(v.ChannelCall("sign_define", errorSign.name, argDict)) != 0 {
+		return fmt.Errorf("sign_define failed")
+	}
+	return nil
+}
+
+// bufferSigns represents a single element in the response from a sign_getplaced() call
+type bufferSigns struct {
+	Signs []struct {
+		Lnum  int    `json:"lnum"`
+		ID    int    `json:"id"`
+		Name  string `json:"name"`
+		Prio  int    `json:"priority"`
+		Group string `json:"group"`
+	} `json:"signs"`
+	BufNr int `json:"bufnr"`
+}
+
+// signGetPlaced returns all signs placed by govim in a specific buffer
+func (v *vimstate) signGetPlaced(buf int) (bufferSigns, error) {
+	argDict := struct {
+		Group string `json:"group"`
+	}{signGroup}
+	resp := v.ParseJSONArgSlice(v.ChannelCall("sign_getplaced", buf, argDict))
+
+	if len(resp) != 1 {
+		// Should never get here since sign_getplaced is called with a single buffer as argument
+		return bufferSigns{}, fmt.Errorf("sign_getplaced failed")
+	}
+
+	var out bufferSigns
+	v.Parse(resp[0], &out)
+	return out, nil
+}
+
+// placeDict is the representation of arguments used in vim's sign_place() and sign_placelist()
+type placeDict struct {
+	Buffer int    `json:"buffer"`          // sign_placelist() only
+	Group  string `json:"group,omitempty"` // sign_placelist() only
+	ID     int    `json:"id,omitempty"`    // sign_placelist() only
+	Lnum   int    `json:"lnum,omitempty"`
+	Name   string `json:"name"` // sign_placelist() only
+	Prio   int    `json:"priority,omitempty"`
+}
+
+// unplaceDict is the representation of arguments used in vim's sign_unplace() and sign_unplacelist()
+type unplaceDict struct {
+	Buffer int    `json:"buffer,omitempty"`
+	Group  string `json:"group,omitempty"` // sign_unplacelist() only
+	ID     int    `json:"id,omitempty"`
+}
+
+// redefineSigns ensures that there is only one govim sign per buffer and line
+// by calculating a difference between current state and the list of quickfix entries
+func (v *vimstate) redefineSigns(fixes []quickfixEntry) error {
+	type bufLine struct {
+		buf  int
+		line int
+	}
+	remove := make(map[bufLine]int) // Value is sign ID, used to unplace duplicates
+	place := make(map[bufLine]int)  // Value is insert order, used to avoid sorting
+
+	// Assume all existing signs should be removed, unless found in quickfix entry list
+	for buf := range v.buffers {
+		placed, _ := v.signGetPlaced(buf)
+		for _, sign := range placed.Signs {
+			bl := bufLine{placed.BufNr, sign.Lnum}
+			if _, exist := remove[bl]; exist {
+				// As each sign isn't tracked individually, we might end up with several
+				// signs on the same line when, for example, a line is removed.
+				// By removing duplicates here we ensure that there is only one
+				// sign per line.
+				v.ChannelCall("sign_unplace", signGroup, unplaceDict{Buffer: bl.buf, ID: sign.ID})
```

Comment:
Any reason you're not using batching here and for the calls that follow?

Reply:
No, not really :). Added batching as well.

Reply:
Oh, turns out goimports can start a batch that isn't terminated before signs are redefined. When using batch to place signs `cmd/govim/testdata/goimports.txt` panics.

I reverted the change for now.

Reply:
This doesn't sound right. There shouldn't be two batches overlapping ever. 

Can you point me to a commit where this panics?